apps/worker/src/app/workflow/services/metric-queue.service.ts

```diff
@@ -0,0 +1,135 @@
+      await this.jobHasCompleted(job);
+    });
+
+    this.bullMqService.worker.on('failed', async (job, error) => {
+      await this.jobHasFailed(job, error);
+    });
+
+    const metricJobExists = Promise.resolve(
+      this.bullMqService.queue.getRepeatableJobs().then(async (job) => {
+        let exists = false;
+        for (const jobElement of job) {
+          if (jobElement.id === 'metric-job') {
+            exists = true;
+          }
+        }
+
+        return exists;
+      })
+    );
+
+    metricJobExists.then((exists) => {
+      Logger.log('metric job exists: ' + exists);
+
+      if (!exists) {
+        Logger.warn("metricJob doesn't exist, creating it");
+        void this.addToQueue('metric-job', {}, '', {
+          jobId: 'metric-job',
+          repeatJobKey: 'metric-job',
+          repeat: {
+            immediately: true,
+            pattern: '* * * * * *',
+          },
+          removeOnFail: true,
+          removeOnComplete: true,
+          attempts: 1,
+        });
+      }
+
+      return true;
+    });
+  }
+
+  public async gracefulShutdown() {
+    // Right now we only want this for testing purposes
+    if (process.env.NODE_ENV === 'test') {
+      await this.bullMqService.queue.drain();
+      await this.bullMqService.worker.close();
+    }
+  }
+
+  private getWorkerOpts(): WorkerOptions {
+    return {
+      ...this.bullConfig,
+      lockDuration: 500,
+      concurrency: 1,
+      settings: {},
+    } as WorkerOptions;
+  }
+
+  private getWorkerProcessor() {
+    return async () => {
+      return await new Promise<void>(async (resolve, reject): Promise<void> => {
+        Logger.log('metric job started');
+
+        for (const queueService of this.token_list) {
+          const metrics = await queueService.bullMqService.getQueueMetrics();
+
+          const completeNumber = metrics.completed.count;
+          const failNumber = metrics.failed.count;
+
+          const successMetric: IMetric = {
+            count: metrics.completed.count,
+            total: completeNumber == 0 ? 0 : sum(metrics.completed.data),
+            min: completeNumber == 0 ? 0 : min(metrics.completed.data),
+            max: completeNumber == 0 ? 0 : max(metrics.completed.data),
+            sumOfSquares: completeNumber == 0 ? 0 : sumOfSquares(metrics.completed.data),
+          };
+
+          const failMetric: IMetric = {
+            count: metrics.failed.count,
+            total: failNumber == 0 ? 0 : sum(metrics.failed.data),
+            min: failNumber == 0 ? 0 : min(metrics.failed.data),
+            max: failNumber == 0 ? 0 : max(metrics.failed.data),
+            sumOfSquares: failNumber == 0 ? 0 : sumOfSquares(metrics.failed.data),
+          };
+
+          if (process.env.NOVU_MANAGED_SERVICE === 'true') {
+            nr.recordMetric(`MetricQueueService/${queueService.name}/completed`, successMetric);
+            nr.recordMetric(`MetricQueueService/${queueService.name}/failed`, failMetric);
```

Comment:
@Cliftonz maybe I'm missing something, but wasn't the most important thing for this is to understand how many jobs we have in flight and pending right now? The completed and failed is quite nice, but I think it lacks the important metrics of how many pending, and how many processing in parallel at the moment. 

Also would be great to know how many are "scheduled" to the future.

Reply:
You are correct, I have updated the code to including getting waiting, active, and delayed jobs.